Join ordering based on input filters

ggevay opened the pull request:
This PR strives to order join inputs in Differential joins and _inside_ each Delta path in a way that inputs that have more selective filters come first. This way, intermediate result sizes can get reduced sooner, so later join steps do less work. This means we have better chances that a filter is applied before some bad stuff (cross joins, new arrangements).

The PR also tweaks Differential ordering to make a smarter choice if the max-minning finds multiple equally good candidates.

(Note that this does no reordering _between_ delta paths.)

As expected from @vmarcos' investigations, this PR has a huge impact in some cases: Speeds up TPC-H Q07 by 10.5x, and Q21 by 8x. This results in a 19% speedup over the sum of all TPC-H queries. (Note that for the speedup in Q07, https://github.com/MaterializeInc/materialize/pull/14725 is also needed.)

(I'm not considering predicates that span over multiple inputs, e.g. `a+b=5` where `a` and `b` come from different inputs. Handling these could be a possible follow-up. For now, it seemed to me that handling these would be much more complicated and they occur less frequently, so the cost-benefit ratio of working on them is not so good.)

### Motivation

  * This PR adds a known-desirable feature: https://github.com/MaterializeInc/materialize/issues/14512

### Tips for reviewer

First of all, don't be alarmed by the large diff; 90+% of it are just json plan changes (edit: I think most of this will go away while addressing comments). Also, many source files have some minor changes just to treat the new field on `MirRelationExpr::Join`.

The main change is in `join_implementation.rs`:
- I added the characteristics of the input filters as a new field on `Characteristics`. This field is used by `Orderer` through the `Ord` of `Characteristics`.
- I also tweaked the code around the max-minning in the Differential ordering, see code comments there.

I needed to add a new field in `MirRelationExpr::Join` for recording the input filter characteristics. The problem that this new field is solving is that `JoinImplementation` lifts away some of the input filters on its first run, so subsequent runs can't see the filters on the inputs anymore. @frankmcsherry, Could you please check that this is ok as a temporary solution? What we could do in the long run, is to lift all filters up all the way, and make them available everywhere as attributes.

I didn't yet add explain printing for the new field:
- The old explain code will soon be deleted.
- I wasn't sure how to add it to the new explain, as the join format there is still in flux.

Note: I'm not looking at the `equivalences` of the `MirRelationExpr::Join`, because at this point in the pipeline, `equivalences` don't contain predicates that can be evaluated in the context of a single input, because `PredicatePushdown` already pushed these down as filters.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
  - I ran TPC-H, see above.
  - We have many join plans in slt tests. I inspected most of the changed plans manually, and didn't find any obviously harmful changes, but found many beneficial changes.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Improve join orders for joins that have filters applied to their inputs, leading to an order of magnitude performance improvement in some cases with highly selective filters.

aalexandrov commented:
## Threshold, Union, Distinct, Negate
New default
```text
EXPLAIN OPTIMIZED PLAN AS TEXT FOR
SELECT a FROM t EXCEPT SELECT b FROM mv
----
Explained Query:
  With
    cte l0 =
      Get materialize.public.t
      Project (#0)
      Distinct group_by=[#0]
    cte l1 =
      Get materialize.public.mv
      Project (#1)
      Distinct group_by=[#0]
      Negate
  Return
    Union
      Get l0
      Get l1
    Threshold

Source materialize.public.mv
  project=(#2, #1)
  map=(dummy)

Used Indexes:
  - materialize.public.t_a_idx
```
vs current default.
```text
EXPLAIN OPTIMIZED PLAN AS TEXT FOR
SELECT a FROM t EXCEPT SELECT b FROM mv
----
Explained Query:
  Threshold
    Union
      Distinct group_by=[#0]
        Project (#0)
          Get materialize.public.t
      Negate
        Distinct group_by=[#0]
          Project (#1)
            Get materialize.public.mv

Source materialize.public.mv
  project=(#2, #1)
  map=(dummy)

Used Indexes:
  - materialize.public.t_a_idx
```

aalexandrov commented:
## `Select` subqueries
New default
```text
EXPLAIN OPTIMIZED PLAN AS TEXT FOR
SELECT (SELECT iv.a FROM iv WHERE iv.b = t.b LIMIT 1), (SELECT mv.a FROM mv WHERE mv.b = t.b LIMIT 1) FROM t
----
Explained Query:
  With
    cte l0 =
      Get materialize.public.t
      Project (#1)
    cte l1 =
      Get l0
      Distinct group_by=[#0]
    cte l2 =
      Get l1
      ArrangeBy keys=[[#0]]
    cte l3 =
      Get materialize.public.iv
      Filter (#1) IS NOT NULL
    cte l4 =
      Join on=(#0 = #2) type=differential
        Get l2
        Get l3
      Project (#0, #1)
      TopK group_by=[#0] limit=1 monotonic=false
    cte l5 =
      Get materialize.public.mv
      Filter (#1) IS NOT NULL
    cte l6 =
      Join on=(#0 = #2) type=differential
        Get l2
        Get l5
      Project (#0, #1)
      TopK group_by=[#0] limit=1 monotonic=false
    cte l7 =
      Get l4
      Project (#0)
      Negate
    cte l8 =
      Union
        Get l7
        Get l1
      Map (null)
    cte l9 =
      Union
        Get l4
        Get l8
      ArrangeBy keys=[[#0]]
    cte l10 =
      Get l6
      Project (#0)
      Negate
    cte l11 =
      Union
        Get l10
        Get l1
      Map (null)
    cte l12 =
      Union
        Get l6
        Get l11
      ArrangeBy keys=[[#0]]
  Return
    Join on=(eq(#0, #1, #3)) type=differential
      Get l0
      Get l9
      Get l12
    Project (#2, #4)

Source materialize.public.mv
  filter=((#1) IS NOT NULL)

Used Indexes:
  - materialize.public.t_a_idx
  - materialize.public.iv_primary_idx
```
vs current default.
```text
EXPLAIN OPTIMIZED PLAN AS TEXT FOR
SELECT * FROM t WHERE EXISTS(SELECT * FROM mv WHERE t.a < mv.a) AND EXISTS(SELECT * FROM mv WHERE t.b > mv.b)
----
Explained Query:
  With
    cte l0 =
      Project (#0, #1)
        Join on=(#0 = #2) type=delta
          ArrangeBy keys=[[#0]]
            Get materialize.public.t
          ArrangeBy keys=[[#0]]
            Distinct group_by=[#0]
              Project (#0)
                Filter (#0 < #1)
                  CrossJoin type=differential
                    ArrangeBy keys=[[]]
                      Distinct group_by=[#0]
                        Project (#0)
                          Get materialize.public.t
                    Project (#0)
                      Get materialize.public.mv
  Return
    Project (#0, #1)
      Join on=(#1 = #2) type=differential
        Get l0
        ArrangeBy keys=[[#0]]
          Distinct group_by=[#0]
            Project (#0)
              Filter (#0 > #1)
                CrossJoin type=differential
                  ArrangeBy keys=[[]]
                    Distinct group_by=[#0]
                      Project (#1)
                        Get l0
                  Project (#1)
                    Get materialize.public.mv

Used Indexes:
  - materialize.public.t_a_idx
```

ggevay commented:
@aalexandrov Are these comments intended for https://github.com/MaterializeInc/materialize/pull/15119?

aalexandrov commented:
Tagging @philip-stoev in case he wants to run a random set of queries grammar to check for bugs. The change introduced here builds upon an existing code in a very straight-forward way. Consequently, the addition of the new heuristic might lead to changes in join orders which I wouldn't necessarily classify as an error (most of the time this should lead to strict improvement, and even if it doesn't there is little that we can do here in the absence of a more accurate cost model).

philip-stoev commented:
Thank you, will work on this tomorrow.

philip-stoev commented:
@ggevay how do you decide which Filter is more selective, in the absence of statistics? The linked issue says "iteral equality, literal inequality, other filter ". Is that still true?

aalexandrov commented:
```rust
#[derive(Eq, PartialEq, Ord, PartialOrd, Debug, Clone, Serialize, Deserialize, Hash, MzReflect)]
pub struct FilterCharacteristics {
    literal_equality: bool,
    like: bool,
    literal_inequality: bool,
    /// Any filter, except ones involving IsNull, because those are too common.
    /// Can be true by itself, or any other field being true also makes this true.
    any_filter: bool,
}
```

I assume that the derived `Ord` on this class uses lexicographic sorting, so for example a predicate that contains a `literal_equality` will be considered more selective than a predicate that has both `like` equality and `literal_inequality`.

aalexandrov reviewed (approved):
LGTM.

Overall, the code is very well structured, but I've noted down a number of minor tweaks that one can do to improve readability for newcomers. 

I think it makes more sense to open a separate PR for that rather than pushing them to this diff, as they are more than a few places.

frankmcsherry reviewed (commented):
I don't have time for a full review at the moment, but I do have a drive-by comment:

Stashing optimization specific metadata in an AST node has a bit of a smell, in that it leaks the abstraction of that optimization into places that are not going to understand them. Other optimizations that move predicates around will need to know to clear the fields, explain plan readers may need to have to learn to understand what they are (just hints afaict?).

Idiomatically, the way we've done this elsewhere is to have the relevant transformation (`join_implementation.rs`) communicate appropriate metadata downwards as part of its analysis, which allows it to reconstruct, use, and discard the information without exposing it to other operators that might invalidate it, etc.

philip-stoev commented:
Item No 1. This join order flipped from one seemingly random order to another. Neither is the syntactic order and I am not sure if any of the predicates can reasonably tilt the decision one way or the other. Is this to be expected? I guess so, but I wanted to confirm.

```
pstoev@Ubuntu-2004-focal-64-minimal:~/randgen$ echo "EXPLAIN  SELECT  FROM lineitem  JOIN orders  ON  TRUE  JOIN customer  ON o_custkey  = c_custkey  ORDER  BY  TRUE   ;" | ~/diff.sh 
EXPLAIN SELECT
  FROM lineitem
  JOIN orders
    ON TRUE
  JOIN customer
    ON o_custkey = c_custkey
 ORDER  BY TRUE ;

--- 6875.out	2022-10-07 10:43:03.234588499 +0200
+++ 16875.out	2022-10-07 10:43:03.274589339 +0200
@@ -1,20 +1,20 @@
 %0 =
 | Get materialize.public.lineitem (####)
 | ArrangeBy ()
 
 %1 =
 | Get materialize.public.orders (####)
 | ArrangeBy (#1)
 
 %2 =
 | Get materialize.public.customer (####)
 | ArrangeBy (#0)
 
 %3 =
 | Join %0 %1 %2 (= #17 #25)
-| | implementation = Differential %2.(#0) %1.(#1) %0.()
+| | implementation = Differential %1.(#1) %2.(#0) %0.()
 | Map true
 | Project (#0..=#24, #17, #26..=#33)
 
 Finish order_by=(#33 asc nulls_last) limit=none offset=0 project=()
```

ggevay commented:
Yes, I would say this change is ok. The only important join ordering consideration that I can see in this query is that the cross join should happen last. If the cross join were somewhere in the middle, then we would have to create an arrangement for the result of the cross join, which would take up a lot of memory.

philip-stoev commented:
Item No2 . In the prior commit, this query had two `IS NOT NULL` filters pushed between the `Get` and the `Arrange`  , which I assume would cause the size of the arrangement to be reduced. This is no longer the case with the current commit:

```
EXPLAIN SELECT *
  FROM lineitem
  JOIN orders
    ON (l_orderkey = o_orderkey)
  JOIN customer
    ON (c_acctbal = o_totalprice)
 WHERE l_extendedprice > 1 ;

--- 6875.out	2022-10-07 10:57:31.356033626 +0200
+++ 16875.out	2022-10-07 10:57:31.400034488 +0200
@@ -1,19 +1,19 @@
 %0 =
 | Get materialize.public.lineitem (####)
 | ArrangeBy (#0)
 
 %1 =
 | Get materialize.public.orders (####)
-| ArrangeBy (#0)
+| Filter (#0) IS NOT NULL, (#3) IS NOT NULL <- HERE THIS WAS THE SITUATION WITH THE PRIOR COMMIT
+| ArrangeBy (#3)
 
 %2 =
 | Get materialize.public.customer (####)
 | Filter (#5) IS NOT NULL
-| ArrangeBy (#5)
 
 %3 =
 | Join %0 %1 %2 (= #0 #16) (= #19 #30)
-| | implementation = Differential %1.(#0) %0.(#0) %2.(#5)
-| Filter (#5 > 1), (#0) IS NOT NULL
+| | implementation = Differential %2 %1.(#3) %0.(#0)
+| Filter (#5 > 1)
 | Project (#0..=#15, #0, #17..=#29, #19, #31, #32)
```

DDL:

```
CREATE TABLE orders (
    o_orderkey       integer,
    o_custkey        integer ,
    o_orderstatus    text ,
    o_totalprice     decimal(15, 2) ,
    o_orderdate      DATE ,
    o_orderpriority  text ,
    o_clerk          text ,
    o_shippriority   integer ,
    o_comment        text 
);

CREATE INDEX pk_orders_orderkey ON orders (o_orderkey);

CREATE INDEX fk_orders_custkey ON orders (o_custkey ASC);

CREATE TABLE lineitem (
    l_orderkey       integer ,
    l_partkey        integer ,
    l_suppkey        integer ,
    l_linenumber     integer ,
    l_quantity       decimal(15, 2) ,
    l_extendedprice  decimal(15, 2) ,
    l_discount       decimal(15, 2) ,
    l_tax            decimal(15, 2) ,
    l_returnflag     text ,
    l_linestatus     text ,
    l_shipdate       date ,
    l_commitdate     date ,
    l_receiptdate    date ,
    l_shipinstruct   text ,
    l_shipmode       text ,
    l_comment        text 
);

CREATE INDEX pk_lineitem_orderkey_linenumber ON lineitem (l_orderkey, l_linenumber);

CREATE INDEX fk_lineitem_orderkey ON lineitem (l_orderkey ASC);
CREATE INDEX fk_lineitem_partkey ON lineitem (l_partkey ASC);
CREATE INDEX fk_lineitem_suppkey ON lineitem (l_suppkey ASC);
CREATE INDEX fk_lineitem_partsuppkey ON lineitem (l_partkey ASC, l_suppkey ASC);
```

philip-stoev commented:
Item No 3. Likely not related to this change per se, but when the plan flipped I noticed that only one of the two inequality conditions in the `WHERE` clause is pushed to a `Get` , but never both. It is just that one or the other is picked depending on the revision:

```
EXPLAIN SELECT *
  FROM lineitem
 RIGHT JOIN orders
    ON (l_orderkey = o_orderkey)
  JOIN customer
    ON (TRUE)
 WHERE o_totalprice > 191
   AND l_partkey != 116 ;

--- 6875.out	2022-10-07 11:11:58.492699347 +0200
+++ 16875.out	2022-10-07 11:11:58.532700270 +0200
@@ -1,18 +1,18 @@
 %0 =
 | Get materialize.public.lineitem (####)
-| ArrangeBy (#0)
+| Filter (#1 != 116), (#0) IS NOT NULL
+| ArrangeBy ()
 
 %1 =
 | Get materialize.public.orders (####)
-| Filter (#3 > 191), (#0) IS NOT NULL
-| ArrangeBy ()
+| ArrangeBy (#0)
 
 %2 =
 | Get materialize.public.customer (####)
 
 %3 =
 | Join %0 %1 %2 (= #0 #16)
-| | implementation = Differential %2 %1.() %0.(#0)
-| Filter (#1 != 116)
+| | implementation = Differential %2 %0.() %1.(#0)
+| Filter (#19 > 191)
 | Project (#0..=#15, #0, #17..=#32)
```

philip-stoev commented:
Item No 4. In this query, an opportunity to filter on `l_linenumber = 7 ` prior to the join has been lost:

```
EXPLAIN SELECT *
  FROM lineitem
  JOIN orders
    ON (l_orderkey = o_orderkey)
  JOIN customer
    ON (TRUE)
 WHERE l_linenumber = 7
   AND o_custkey > 45 ;

--- 6875.out	2022-10-07 11:21:28.157225944 +0200
+++ 16875.out	2022-10-07 11:21:28.205226964 +0200
@@ -1,18 +1,18 @@
 %0 =
 | Get materialize.public.lineitem (####)
-| ArrangeBy (#0)
+| Filter (#3 = 7), (#0) IS NOT NULL
+| ArrangeBy ()
 
 %1 =
 | Get materialize.public.orders (####)
 | ArrangeBy (#0)
 
 %2 =
 | Get materialize.public.customer (####)
-| ArrangeBy ()
 
 %3 =
 | Join %0 %1 %2 (= #0 #16)
-| | implementation = Differential %1.(#0) %0.(#0) %2.()
-| Filter (#3 = 7), (#17 > 45), (#0) IS NOT NULL
+| | implementation = Differential %2 %0.() %1.(#0)
+| Filter (#17 > 45)
 | Project (#0..=#15, #0, #17..=#32)
```

philip-stoev commented:
Item No 5. Very weirdly, in this commit this obviously impossible query is detected as such, but the prior commit has a full blown plan. There is nothing special in the impossible `WHERE` clause so I wonder why it was not detected as such immediately.

```
EXPLAIN OPTIMIZED PLAN AS TEXT
   FOR SELECT *
  FROM lineitem
  LEFT JOIN orders
    ON (l_orderkey = o_orderkey)
  JOIN customer
    ON (TRUE)
 WHERE o_custkey = 58
   AND o_custkey = 158 ;

--- 6875.out	2022-10-07 11:29:18.099070044 +0200
+++ 16875.out	2022-10-07 11:29:18.143070956 +0200
@@ -1,3 +1,16 @@
-Explained Query (fast path):
-  Constant
+Explained Query:
+  Project (#0..=#15, #0, #17..=#32)
+    Filter (#17 = 58) AND (#17 = 158)
+      Join on=(#0 = #16) type=differential
+        ArrangeBy keys=[[]]
+          Filter (#0) IS NOT NULL
+            Get materialize.public.lineitem
+        ArrangeBy keys=[[#0]]
+          Get materialize.public.orders
+        Get materialize.public.customer
+
+Used Indexes:
+  - materialize.public.pk_customer_custkey
+  - materialize.public.pk_orders_orderkey
+  - materialize.public.pk_lineitem_orderkey_linenumber
```

ggevay commented:
Item No2. In the new order we are re-using the existing primary key arrangement on `orders`, and when we re-use an arrangement then we can't have a filter between the `Get` and the `ArrangeBy` (because in this case the `Get` and the  `ArrangeBy` together just mean re-using the existing arrangement).

Btw. the new explain output shows which indexes are being used:
```
 EXPLAIN optimized plan WITH(join_impls) as text for SELECT *
  FROM lineitem
  JOIN orders
    ON (l_orderkey = o_orderkey)
  JOIN customer
    ON (c_acctbal = o_totalprice)
 WHERE l_extendedprice > 1 ;
```

Before:
```
                      Optimized Plan                      
----------------------------------------------------------
 Explained Query:                                        +
   Project (#0..=#15, #0, #17..=#29, #19, #31, #32)      +
     Filter (#5 > 1)                                     +
       Join on=(#0 = #16 AND #19 = #30) type=differential+
         implementation                                  +
           materialize.public.customer » %1[#3] » %0[#0] +
         ArrangeBy keys=[[#0]]                           +
           Get materialize.public.lineitem               +
         ArrangeBy keys=[[#3]]                           +
           Filter (#0) IS NOT NULL AND (#3) IS NOT NULL  +
             Get materialize.public.orders               +
         Get materialize.public.customer                 +
                                                         +
 Used Indexes:                                           +
   - materialize.public.pk_orders_orderkey               +
   - materialize.public.fk_lineitem_orderkey             +
   - materialize.public.pk_customer_custkey              +
```

After:
```
                      Optimized Plan                      
----------------------------------------------------------
 Explained Query:                                        +
   Project (#0..=#15, #0, #17..=#29, #19, #31, #32)      +
     Filter (#5 > 1) AND (#0) IS NOT NULL                +
       Join on=(#0 = #16 AND #19 = #30) type=differential+
         implementation                                  +
           %1 » %0[#0] » %2[#5]                          +
         ArrangeBy keys=[[#0]]                           +
           Get materialize.public.lineitem               +
         ArrangeBy keys=[[#0]]                           +
           Get materialize.public.orders                 +
         ArrangeBy keys=[[#5]]                           +
           Get materialize.public.customer               +
                                                         +
 Used Indexes:                                           +
   - materialize.public.pk_orders_orderkey               +
   - materialize.public.fk_lineitem_orderkey             +
   - materialize.public.pk_customer_custkey              +
```

But unfortunately the output is quite misleading here, because it conflates using an index to perform a join stage and just an index scan: both the "before" and "after" plans say we are using `pk_orders_orderkey`, but in the "before" case it means just performing the `Get` that is before the `Filter` by an index scan, and in the "after" case it means performing the `» %0[#0]` part of the join using the index. I opened an issue about this: https://github.com/MaterializeInc/materialize/issues/15242

ggevay commented:
Re Item No 3. and 4.: These changes are both because of re-using existing arrangements.

In 4., the old plan might be better: we should apply the filter before the costly cross join. We could add a heuristic to don't re-use an existing arrangement before a cross join if this would mean lifting a filter. However, it's not clear even to my human eyes whether the old plan really is better, because the new plan has the cross join later, which means that the other filter was applied before.

So, the better long-term solution to this dilemma of whether to re-use an arrangement or filter the input, would be to gather run-time statistics on filter selectivities and join sizes, and do cost-based join optimization.

ggevay commented:
Re Item No 5.: We currently detect something like `o_custkey = 58 AND o_custkey = 158` only in `CanonicalizeMfp`, and only if it's right on top of a `Get` (because I added this as part of the literal constraints detection). Here, it's not being detected in the "before" case because of `JoinImplementation` lifting the filter away from the Get, which is happening (again) due to different arrangement re-use. In the long run, we should have a detection for such impossible predicates in `MirScalarExpr::reduce`, and then we'll be able to detect these anywhere in the plan.

Btw. this prompted me to think about adding another run of `CanonicalizeMfp` to before `JoinImplementation`, and then we would detect this impossible filter before it's lifted away from the `Get`. And having `CanonicalizeMfp` before `JoinImplementation` would also have other advantages: we would detect literal constraints on join inputs and not lift them away if an arrangement usable for the literal constraint is present.

Thank you very much for the thorough testing @philip-stoev!

philip-stoev commented:
Item No 8. More generally, I can not get the more selective filter to be ordered first:

Those two queries:

```
explain select * from t1 , t2, t3 where t2.f2 = 'a' and t1.f2 > 'b';
explain select * from t1 , t2, t3 where t1.f2 = 'a' and t2.f2 > 'b';
```

have the same join order. Again, if this is supposed to change in this PR, an example would be appreciated.

philip-stoev commented:
Item No 9. I think I am getting somewhere, please bear with me :-( . For this dataset:

```
CREATE TABLE t1 (f1 TEXT, f2 TEXT);
CREATE INDEX i1 ON t1(f1);
CREATE TABLE t2 (f1 TEXT, f2 TEXT);
CREATE INDEX i2 ON t2(f1);
CREATE TABLE t3 (f1 TEXT, f2 TEXT);
CREATE INDEX i3 ON t3(f1);
CREATE TABLE t4 (f1 TEXT, f2 TEXT);
CREATE INDEX i4 ON t4(f1);
```

Those two queries have the same plan even though the two filters have different potential selectivity:

```
explain select * from t1 , t2, t3, t4 where t1.f1 = t2.f1 and t2.f1 = t3.f1 and t3.f1 = t4.f1 and t3.f2 not like 'a' and t4.f2 like 'a';
explain select * from t1 , t2, t3, t4 where t1.f1 = t2.f1 and t2.f1 = t3.f1 and t3.f1 = t4.f1 and t4.f2 not like 'a' and t3.f2 like 'a';
```

I am not sure if this falls under this PR, but, in the absence of statistics, `LIKE` needs to be "costed" as a `BETWEEN`, that is, an OR of two inequalities, that is one expects to match some small portion of the table. `NOT LIKE` on the other hand is "costed" as an inequality, that is, one expects most of the table to match.

ggevay commented:
Item No 6., 7., 8.: The current main [assumes](https://github.com/MaterializeInc/materialize/blob/83bc683394dfcb20dff73075e402310469501b1f/src/transform/src/join_implementation.rs#L377) that in case of a Differential join the characteristics of the first input don't matter. I carried over this assumption to this PR. This might cease to be true in the future after fixing some other issues that we have around Differential joins: https://github.com/MaterializeInc/materialize/issues/14059, https://github.com/MaterializeInc/materialize/issues/15243

ggevay commented:
Item No 9.:

Good point about handling LIKE and NOT LIKE differently! I'll add a case for this.

For whether LIKE is similar to an inequality or between, I'm not sure. My gut feeling is that LIKE is more restrictive then an inequality or a between, but NOT LIKE is less restrictive. I vaguely remember reading it somewhere that, in the absence of statistics, database systems often assume 1/3 selectivity for an inequality. This makes sense if we assume uniform distribution of data, and then a predicate `column > x` with a uniform choice of `x` in the domain of the data. For a LIKE, a uniform distribution of the data would mean 1/26 selectivity even for `LIKE 'a%'` assuming random strings of lower-case letters.

philip-stoev commented:
Item No 10. `IS NULL` and `IS NOT NULL` seem to be disregarded completely:

```
explain select * from t1 , t2, t3, t4 where t1.f1 = t2.f1 and t2.f1 = t3.f1 and t3.f1 = t4.f1 and (t4.f2 is not null and t3.f2 is null);
```

This produces a delta plan where t1 and t2 are first.

I think IS NULL should be considered quite restrictive and `NOT NULL` should be considered to match almost everything (?) Man, how we need histograms! If you can think of a rule to incorporate those, please proceed as you see fit.

philip-stoev commented:
Item No 11. Back to `BETWEEN`, it seems that `BETWEEN` is not considered more restrictive then simple inequality.

```
 explain select * from t1 , t2, t3, t4 where t1.f1 = t2.f1 and t2.f1 = t3.f1 and t3.f1 = t4.f1 and (t4.f2 between 'a' and 'b' and t3.f2 > 'a');
```

This does not cause t4 to be ordered before t3.

ggevay commented:
Item No 10.: There is way too many `IS NOT NULL` to be useful, so I'm explicitly ignoring them when determining the "is there any filter present" characteristic. However, `IS NULL` might be important, so maybe I'll add a case for that.

ggevay commented:
Item No. 11.: There is only so much we can do in the absence of histograms... Maybe I'll change the heuristics to not just have a present/not-present boolean for certain filters, but count how many of them we have.

philip-stoev commented:
And a final one, I promise. The tables in this query retain the syntactic join order:

```
explain select * from t1 , t2, t3, t4 where t1.f1 = t2.f1 and t2.f1 = t3.f1 and t3.f1 = t4.f1 and (t4.f2 is null or t4.f2 != 'a');
```

It is truly the cornerest of corner cases, seems to require a combination of `IS NULL` and inequality .

ggevay commented:
For the last one (`t4.f2 is null or t4.f2 != 'a'`), this is currently being disregarded from the `any_filter` category because of the `IsNull` subexpression (which was intended to disregard `IS NOT NULL`). This will be fixed when I separate the handling of `IS NULL` and `IS NOT NULL`.

philip-stoev reviewed (approved):
Ok so before I overstay my welcome in this PR:
- no wrong results from randomized testing
- I examined several dozen changed plans in TPC-*-like generated queries and the differences all fall into the Items initially reported
- I then did some manual experimentation. Unfortunately, I do not have the time to build a full-blown SLT file out of the queries raised as comments, but if you think some are useful, please consider adding a dedicated SLT file for this optimization, rather than relying on query plans in other tests that have changed

ggevay commented:
Thanks a lot @philip-stoev! To sum up my plans for addressing your comments:
- (The filters moving around are due to different arrangement re-use.)
- (The characteristics of the first input of Differential joins is currently being disregarded. This might change in the future after some other issues are fixed.)
- I opened 2 new issues (not in the scope of this PR) while investigating your comments:
https://github.com/MaterializeInc/materialize/issues/15242,
https://github.com/MaterializeInc/materialize/issues/15243
- I'll improve the heuristics in `filter_characteristics`:
  - Separate the handling of detected filters from their NOTted versions.
  - Maybe change from boolean present/not-present characteristics to instead count the number of filters having the characteristic.
- I'll add some cases in the existing `joins.slt`.

ggevay commented:
I'm trying to finalize this PR. Somehow, a 7000-line diff appeared in `test/sqllogictest/explain/decorrelated_plan_as_json.slt`. I think this must be some weird rebase fail, because I didn't explicitly change anything there (the join ordering shouldn't have any effect on `EXPLAIN DECORRELATED PLAN`). It seems that the diff is caused by `WITH(raw)` having disappeared from the explain commands somehow.

@aalexandrov,
What does `WITH(raw)` mean? Should it be documented [here](https://materialize.com/docs/sql/explain/#output-modifiers)?
Should I simply revert the file, so that my PR won't be touching it?

aalexandrov commented:
`raw` is basically a guard of doing any transformations of the tree or using virtual syntax before printing it, possibly at the expense of forcefully disabling other conflicting flags. I think that for `explain_~_as_json.slt` files this should be always part of the `EXPLAIN` so the JSON reflects the underlying expression as closely as possible.

> Should I simply revert the file, so that my PR won't be touching it?
Yes, you can revert and then again rewrite with `--rewrite-results` so you apply only the changes that are coming as part of this PR (but keep the `WITH(raw)` when rewriting).

ggevay commented:
Thanks, @aalexandrov, I reverted that file.

Now the PR is ready for review again. The 3 commits:
- The 1st commit adds `visit_pre_with_context` as discussed here: https://github.com/MaterializeInc/materialize/pull/15307
- The 2nd commit does a minor refactor of `PredicatePushdown`, so that `JoinImplementation` can access some of the pushdown logic for determining which filters belong to which inputs even after an earlier run of `JoinImplementation` undos the work of `PredicatePushdown`.
- The 3rd commit is the main work.

I addressed Philip's comments as outlined [above](https://github.com/MaterializeInc/materialize/pull/15120#issuecomment-1271576119). `FilterCharacteristics` is smarter now, but still far from perfect. I think further meaningful improvement will be possible only when we will be gathering statistics.

I removed the extra field that I was adding to `MirRelationExpr::Join` in the original version of the PR, thus addressing @frankmcsherry's [comment](https://github.com/MaterializeInc/materialize/pull/15120#pullrequestreview-1133122046). The information that I was saving in the extra field is now reconstructed by calling some of the `PredicatePushdown` logic (using the MFP that is above the join).

I haven't yet changed the EXPLAIN output for joins (adding FilterCharacteristics and other info). I'll do this in a follow-up PR, because this PR is already quite big.

aalexandrov commented:
@ggevay thanks, I'll review it tomorrow.

aalexandrov commented on src/expr/src/scalar/mod.rs:
```diff
@@ -1961,6 +1982,129 @@ impl VisitChildren<Self> for MirScalarExpr {
+                        )
+                },
+                &mut |not_in_parent_chain, expr| {
+                    if !not_in_parent_chain {
```
Can you put a comment here explaining why you guard the logic below in `!not_in_parent_chain` for posteriy (it's not immediately obvious).

aalexandrov reviewed (approved):
I left some minor comments (nits and non-PR related musings), otherwise LGTM.

ggevay commented on src/transform/src/join_implementation.rs:
```diff
@@ -126,21 +135,93 @@ impl JoinImplementation {
+                        ..
+                    } = input
+                    {
+                        let (mfp, input) =
```
Yeah, I'm becoming more and more convinced nowadays that we should move all these arrangement things to LIR. Having these ArrangeBys in MIR is not so useful, because it's very hard to guess
- whether it's forming an arrangement right there or just a marker that that stuff is already arranged;
- whether the arrangement will be scanned or used in a lookup (https://github.com/MaterializeInc/materialize/issues/15242).

And then EXPLAINing indexes would become the responsibility of EXPLAIN PHYSICAL PLAN.

aalexandrov replied:
AFAIU the main obstacle at the moment is the fact that certain MIR optimizations (such as `JoinOrdering`) impose a constraint that an input should be arranged, and if the input is consumed more than once in a `Join` node, a first run might come up with a differential plan that is guaranteed to ad-hoc create all arrangements that were missing to pick a delta plan.

> And then EXPLAINing indexes would become the responsibility of EXPLAIN PHYSICAL PLAN.

I think we can (and probably should) do this even now, and it will be much more robust if we do this kind of inference on top of LIR.

ggevay commented on src/transform/src/predicate_pushdown.rs:
```diff
@@ -754,6 +695,75 @@ impl PredicatePushdown {
+        let new_inputs = inputs
+            .drain(..)
+            .zip(push_downs)
+            .enumerate()
```
Done

ggevay commented on src/expr/src/scalar/mod.rs:
```diff
@@ -1961,6 +1982,129 @@ impl VisitChildren<Self> for MirScalarExpr {
+/// ordering for join inputs. Bigger is better, i.e., will tend to come to the front of the inputs.
+#[derive(Eq, PartialEq, Ord, PartialOrd, Debug, Clone, Serialize, Deserialize, Hash, MzReflect)]
+pub struct FilterCharacteristics {
+    literal_equality: bool,
```
`literal_inequality` is `usize` to handle `BETWEEN`.

For `literal_equality`, you are right that `#0 = 0 AND #1 = 2` could benefit from `literal_equality` also being `usize`, but the issue is that literal equalities often appear in ORs, e.g. an IN list is translated to `#0 = 5 OR #0 = 7`, in which case more of them means less selectivity. Of course, this can also happen with literal inequalities (that they appear in an OR), but it's maybe less typical for those. These things are very heuristic, sorry for that! We could tweak these characteristics endlessly, but I think the long-term solution will be to measure filter selectivity at run time (or at least gather statistics by sampling stuff that is in persist).

ggevay commented on src/expr/src/scalar/mod.rs:
```diff
@@ -1961,6 +1982,129 @@ impl VisitChildren<Self> for MirScalarExpr {
+                        )
+                },
+                &mut |not_in_parent_chain, expr| {
+                    if !not_in_parent_chain {
```
I extended the FilterCharacteristics field docstring
